azure/scope/cluster.go

```diff
@@ -392,8 +406,15 @@ func (s *ClusterScope) Vnet() *infrav1.VnetSpec {
 }
 
 // IsVnetManaged returns true if the vnet is managed.
-func (s *ClusterScope) IsVnetManaged() bool {
-	return s.Vnet().ID == "" || s.Vnet().Tags.HasOwned(s.ClusterName())
+func (s *ClusterScope) IsVnetManaged(ctx context.Context) (bool, error) {
+	var err error
+	if s.cache.IsVnetManaged == nil {
+		vnetSvc := virtualnetworks.New(s)
```

Comment:
@shysank PTAL. I had to go back on our conversation in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1610#discussion_r698808313 because it doesn't work for Delete() as we call services in the opposite order as Reconcile() (eg. the subnets need to be deleted before the vnet is deleted) so this means the vnet service isn't the first one to be reconciled which means we have to do this before we get to the vnet service. I'm not sure how to work around that constraint without calling the virtual network from either the scope or other services (eg. from the subnet service). Thoughts?

Reply:
cc @devigned as well

Reply:
can we think of `IsManaged` as something that identifies whether that resource is managed by capz instead of checking `IsVnetManaged`? Every Service can implement `IsManaged` behaviour, and internally a service can depend on another service or another service's client to determine whether it is managed or not. The cache will be shared among all services. In this way, we can make clear abstractions. Concretely: `SubnetSvc.IsManaged()` will check if vnet is managed either through `virtualNetworksClient` (we can write a helper if it is used in more than one place) or by exposing `virtualNetworksSvc.IsManaged`, and updates the cache through scope which can be used by subsequent services. wdyt?

Reply:
let me try some things once #1874 is merged and I can rebase on top of it